Bridge.Translator error while trying to close the project file under Linux and OS X on Mono + msbuild

Bridge Translator's OpenProject() and CloseProject() calls to `msbuild` that attempts to load and unload projects fails during Bridge Build time.

### Steps To Reproduce

- Reproducible at least on VS for mac (possibly on linux as well)
- Install Visual Studio for Mac
- Create a new project/solution under .NET > Library (ClassLibrary)
- Install Bridge 17.1.0 NuGet packages

#### Scenario 1: all from VS4Mac
- Try to build the project from VS4Mac (you may close and re-open the solution to ensure bridge settings kicks in)

#### Scenario 2: project created from VS4Mac, built from Bridge CLI
- Open a console
- install bridge 17.1.0 from homebrew if not installed (`brew install bridgedotnet/cli/bridge`)
  (update it if already exists: `brew update bridge`)
- Change directory to the project's root (not solution's)
- run `bridge build`

### Expected Result

Build is successful in either scenario.

### Actual Result

#### Scenario 1:
Error during Bridge build:
`The project file "<project path>" cannot be unloaded because at least one project "<same project path>" is still loaded which references that project XML.`
From: Microsoft.Build package's `UnloadProject()` call made by `Translator.CloseProject()`.

#### Scenario 2:
Error during bridge build:
`[ERROR] FATAL UNHANDLED EXCEPTION: Could not load type of field from assembly `System.Threading.Tasks.Dataflow`, initiated from `Translator.OpenProject()`. (message not exactly what shows up on the mac output for simplicity).

--- a/.build/specs/Bridge.Min.nuspec
+++ b/.build/specs/Bridge.Min.nuspec
@@ -73,6 +73,7 @@
     <file src="../../Compiler/Translator/bin/Release/Microsoft.Build.dll" target="tools/Microsoft.Build.dll" />
     <file src="../../Compiler/Translator/bin/Release/Microsoft.Build.Framework.dll" target="tools/Microsoft.Build.Framework.dll" />
     <file src="../../Compiler/Translator/bin/Release/Microsoft.Build.Utilities.Core.dll" target="tools/Microsoft.Build.Utilities.Core.dll" />
+    <file src="../../Compiler/Translator/bin/Release/System.Threading.Tasks.Dataflow.dll" target="tools/System.Threading.Tasks.Dataflow.dll" />
     <file src="../../Compiler/Translator/bin/Release/System.Runtime.InteropServices.RuntimeInformation.dll" target="tools/System.Runtime.InteropServices.RuntimeInformation.dll" />
   </files>
 </package>

--- a/Compiler/Translator/Bridge.Translator.csproj
+++ b/Compiler/Translator/Bridge.Translator.csproj
@@ -54,15 +54,15 @@
       <HintPath>..\..\packages\Bridge.NRefactory.5.5.9\lib\net40\ICSharpCode.NRefactory.CSharp.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Build.15.6.85\lib\net46\Microsoft.Build.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Build.15.7.179\lib\net46\Microsoft.Build.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.Build.Framework, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Build.Framework.15.6.85\lib\net46\Microsoft.Build.Framework.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Build.Framework.15.7.179\lib\net46\Microsoft.Build.Framework.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.Build.Utilities.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Build.Utilities.Core.15.6.85\lib\net46\Microsoft.Build.Utilities.Core.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Build.Utilities.Core.15.7.179\lib\net46\Microsoft.Build.Utilities.Core.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.CodeAnalysis, Version=2.7.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
@@ -72,6 +72,10 @@
       <HintPath>..\..\packages\Microsoft.CodeAnalysis.CSharp.2.7.0\lib\netstandard1.3\Microsoft.CodeAnalysis.CSharp.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.CSharp" />
+    <Reference Include="Microsoft.VisualStudio.Setup.Configuration.Interop, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\Microsoft.VisualStudio.Setup.Configuration.Interop.1.16.30\lib\net35\Microsoft.VisualStudio.Setup.Configuration.Interop.dll</HintPath>
+      <EmbedInteropTypes>True</EmbedInteropTypes>
+    </Reference>
     <Reference Include="Mono.Cecil, Version=0.9.6.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756, processorArchitecture=MSIL">
       <HintPath>..\..\packages\Mono.Cecil.0.9.6.4\lib\net45\Mono.Cecil.dll</HintPath>
       <Private>True</Private>
@@ -125,6 +129,7 @@
       <HintPath>..\..\packages\System.Composition.TypedParts.1.0.31\lib\portable-net45+win8+wp8+wpa81\System.Composition.TypedParts.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="System.Configuration" />
     <Reference Include="System.Console, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
       <HintPath>..\..\packages\System.Console.4.3.0\lib\net46\System.Console.dll</HintPath>
       <Private>True</Private>
@@ -182,6 +187,10 @@
       <HintPath>..\..\packages\System.Text.Encoding.CodePages.4.3.0\lib\net46\System.Text.Encoding.CodePages.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="System.Threading.Tasks.Dataflow, Version=4.5.24.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\System.Threading.Tasks.Dataflow.4.5.24\lib\portable-net45+win8+wpa81\System.Threading.Tasks.Dataflow.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
     <Reference Include="System.Threading.Thread, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
       <HintPath>..\..\packages\System.Threading.Thread.4.3.0\lib\net46\System.Threading.Thread.dll</HintPath>
       <Private>True</Private>
@@ -190,6 +199,7 @@
       <HintPath>..\..\packages\System.ValueTuple.4.3.1\lib\netstandard1.0\System.ValueTuple.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="System.Xaml" />
     <Reference Include="System.Xml.Linq" />
     <Reference Include="System.Data.DataSetExtensions" />
     <Reference Include="System.Data" />

--- a/Compiler/Translator/Translator/Translator.ProjectReader.cs
+++ b/Compiler/Translator/Translator/Translator.ProjectReader.cs
@@ -39,19 +39,7 @@ namespace Bridge.Translator
 
         internal static void CloseProject(Project project)
         {
-            bool isOnMono = Translator.IsRunningOnMono();
-            if (isOnMono)
-            {
-                // This UnloadProject overload should be used if the project created by new Project(XmlReader.Create(this.Location)...)
-                // Otherwise it does NOT work either on Windows or Linux
-                project.ProjectCollection.UnloadProject(project.Xml);
-            }
-            else
-            {
-                // This UnloadProject overload should be used if the project created by new Project(this.Location...)
-                // Otherwise it does NOT work either on Windows or Linux
-                project.ProjectCollection.UnloadProject(project);
-            }
+            project.ProjectCollection.UnloadProject(project);
         }
 
         internal virtual void EnsureProjectProperties()

--- a/Compiler/Translator/packages.config
+++ b/Compiler/Translator/packages.config
@@ -2,13 +2,14 @@
 <packages>
   <package id="AjaxMin" version="5.14.5506.26202" targetFramework="net45" />
   <package id="Bridge.NRefactory" version="5.5.9" targetFramework="net461" />
-  <package id="Microsoft.Build" version="15.6.85" targetFramework="net461" />
-  <package id="Microsoft.Build.Framework" version="15.6.85" targetFramework="net461" />
-  <package id="Microsoft.Build.Utilities.Core" version="15.6.85" targetFramework="net461" />
+  <package id="Microsoft.Build" version="15.7.179" targetFramework="net461" />
+  <package id="Microsoft.Build.Framework" version="15.7.179" targetFramework="net461" />
+  <package id="Microsoft.Build.Utilities.Core" version="15.7.179" targetFramework="net461" />
   <package id="Microsoft.CodeAnalysis.Analyzers" version="2.6.0" targetFramework="net461" developmentDependency="true" />
   <package id="Microsoft.CodeAnalysis.Common" version="2.7.0" targetFramework="net461" />
   <package id="Microsoft.CodeAnalysis.CSharp" version="2.7.0" targetFramework="net461" />
   <package id="Microsoft.Composition" version="1.0.31" targetFramework="net461" />
+  <package id="Microsoft.VisualStudio.Setup.Configuration.Interop" version="1.16.30" targetFramework="net461" developmentDependency="true" />
   <package id="Mono.Cecil" version="0.9.6.4" targetFramework="net45" />
   <package id="Newtonsoft.Json" version="11.0.2" targetFramework="net461" />
   <package id="Object.NET.Utilities" version="2.5.0" targetFramework="net45" />
@@ -53,6 +54,7 @@
   <package id="System.Text.Encoding.Extensions" version="4.3.0" targetFramework="net461" />
   <package id="System.Threading" version="4.3.0" targetFramework="net461" />
   <package id="System.Threading.Tasks" version="4.3.0" targetFramework="net461" />
+  <package id="System.Threading.Tasks.Dataflow" version="4.5.24" targetFramework="net461" />
   <package id="System.Threading.Tasks.Parallel" version="4.3.0" targetFramework="net461" />
   <package id="System.Threading.Thread" version="4.3.0" targetFramework="net461" />
   <package id="System.ValueTuple" version="4.3.1" targetFramework="net461" />

--- a/Compiler/TranslatorTests/Bridge.Translator.Tests.csproj
+++ b/Compiler/TranslatorTests/Bridge.Translator.Tests.csproj
@@ -53,16 +53,20 @@
       <HintPath>..\..\packages\Bridge.NRefactory.5.5.9\lib\net40\ICSharpCode.NRefactory.CSharp.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Build.15.6.85\lib\net46\Microsoft.Build.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Build.15.7.179\lib\net46\Microsoft.Build.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.Build.Framework, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Build.Framework.15.6.85\lib\net46\Microsoft.Build.Framework.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Build.Framework.15.7.179\lib\net46\Microsoft.Build.Framework.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.Build.Utilities.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Build.Utilities.Core.15.6.85\lib\net46\Microsoft.Build.Utilities.Core.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Build.Utilities.Core.15.7.179\lib\net46\Microsoft.Build.Utilities.Core.dll</HintPath>
       <Private>True</Private>
+    </Reference>
+    <Reference Include="Microsoft.VisualStudio.Setup.Configuration.Interop, Version=1.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\Microsoft.VisualStudio.Setup.Configuration.Interop.1.16.30\lib\net35\Microsoft.VisualStudio.Setup.Configuration.Interop.dll</HintPath>
+      <EmbedInteropTypes>True</EmbedInteropTypes>
     </Reference>
     <Reference Include="Mono.Cecil, Version=0.9.6.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756, processorArchitecture=MSIL">
       <HintPath>..\..\packages\Mono.Cecil.0.9.6.4\lib\net45\Mono.Cecil.dll</HintPath>
@@ -89,6 +93,7 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="System.ComponentModel.Composition" />
+    <Reference Include="System.Configuration" />
     <Reference Include="System.Core" />
     <Reference Include="System.Diagnostics.Process, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
       <HintPath>..\..\packages\System.Diagnostics.Process.4.1.0\lib\net461\System.Diagnostics.Process.dll</HintPath>
@@ -117,6 +122,10 @@
       <HintPath>..\..\packages\System.Security.Cryptography.Primitives.4.3.0\lib\net46\System.Security.Cryptography.Primitives.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="System.Threading.Tasks.Dataflow, Version=4.5.24.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\System.Threading.Tasks.Dataflow.4.5.24\lib\portable-net45+win8+wpa81\System.Threading.Tasks.Dataflow.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
     <Reference Include="System.Threading.Thread, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
       <HintPath>..\..\packages\System.Threading.Thread.4.0.0\lib\net46\System.Threading.Thread.dll</HintPath>
     </Reference>
@@ -125,6 +134,7 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="System.Web" />
+    <Reference Include="System.Xaml" />
     <Reference Include="System.Xml.Linq" />
     <Reference Include="System.Data.DataSetExtensions" />
     <Reference Include="System.Data" />

--- a/Compiler/TranslatorTests/packages.config
+++ b/Compiler/TranslatorTests/packages.config
@@ -1,9 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="Bridge.NRefactory" version="5.5.9" targetFramework="net461" />
-  <package id="Microsoft.Build" version="15.6.85" targetFramework="net461" />
-  <package id="Microsoft.Build.Framework" version="15.6.85" targetFramework="net461" />
-  <package id="Microsoft.Build.Utilities.Core" version="15.6.85" targetFramework="net461" />
+  <package id="Microsoft.Build" version="15.7.179" targetFramework="net461" />
+  <package id="Microsoft.Build.Framework" version="15.7.179" targetFramework="net461" />
+  <package id="Microsoft.Build.Utilities.Core" version="15.7.179" targetFramework="net461" />
+  <package id="Microsoft.VisualStudio.Setup.Configuration.Interop" version="1.16.30" targetFramework="net461" developmentDependency="true" />
   <package id="Mono.Cecil" version="0.9.6.4" targetFramework="net45" />
   <package id="Newtonsoft.Json" version="11.0.2" targetFramework="net461" />
   <package id="Node.js" version="5.3.0" targetFramework="net45" />

